tgfloofbot/plugins/administration/commands.py

```diff
@@ -0,0 +1,117 @@
+from typing import List, Optional
+
+import pydantic
+import telegram
+import datetime
+from datetime import timezone
+
+from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update, ParseMode
+from telegram.ext import CallbackContext, CallbackQueryHandler
+
+from ... import loader
+
+from ...client import TGFloofbotClient
+from ...helpers import em
+from ...logger import LOG
+
+from . import models
+from . import exceptions
+
+@loader.custom
+def administration_custom(client: TGFloofbotClient):
+
+    admin_group_id = client.config.main_group
+
+    def _resolve_admin_group_id(update):
+        try:
+            admin_group = client.updater.bot.get_chat(chat_id=admin_group_id)
+        except telegram.error.TelegramError:
+            LOG.critical("Unable to resolve the group ID set in the config! Defaulting to context group ID.")
+            admin_group = update.effective_chat
+
+        return admin_group
+
+    class WarnCommandArgs(pydantic.BaseModel):
+        bad_user: Optional[str] = pydantic.Field(
+            None, description="A raw user ID")
+        warn_message: Optional[str] = pydantic.Field(
+            None, description="The warning reason")
+
+    class UsernoteCommandArgs(pydantic.BaseModel):
+        bad_user: Optional[str] = pydantic.Field(
+            None, description="A raw user ID")
+        warn_message: Optional[str] = pydantic.Field(
+            None, description="The note")
+
+    @loader.command(name="warn", help="warn a user", admin=True)
+    def warn_command(
+        client: TGFloofbotClient,
+        update: Update,
+        context: CallbackContext,
+        args: WarnCommandArgs
+    ) -> None:
+        warn_helper(client, update, context, args, False)
+        return
+
+    @loader.command(name="note", help="adds a moderation note for a user", admin=True)
+    def usernote_command(
+        client: TGFloofbotClient,
+        update: Update,
+        context: CallbackContext,
+        args: UsernoteCommandArgs
+    ) -> None:
+        warn_helper(client, update, context, args, True)
+        return
+
+    def warn_helper(client, update, context, args, is_note):
+        user = update.effective_user
+        chat = update.effective_chat
+        admin_chat = _resolve_admin_group_id(Update)
+
+        if admin_chat.type != "group":  # group only command
+            context.bot.send_message(chat_id=chat, text="This command can only be used for groups.")
+            return
+        
+        if args.bad_user:
+            try:
+                bad_user = client.updater.bot.get_chat_member(chat_id=admin_chat.id, user_id=args.bad_user).user  # type: ignore
+                if args.warn_message:
+                    reason = args.warn_message
+                else:
+                    reason = "No reason provided."
```

Comment:
This can also be removed if the change involving making `args.warn_message` required is made.

If it has to be kept, convention states that it should be moved outside the `try`/`except` block, as it is impossible for this to raise a `telegram.error.BadRequest`.

Furthermore, if you want to keep a default reason, you can instead specify it in `WarnCommandArgs` and `UsernoteCommandArgs` with `warn_message: str = pydantic.Field("No reason provided.", description="The warning reason")` and remove these 4 lines entirely.